public/scripts/models/image-upload.js

```diff
@@ -45,23 +39,75 @@ class ImageUpload {
     });
     this.file = file;
     this.maxFileSizeByte = maxFileSizeByte;
-    if (!this.supportedFileTypes.includes(file.type)) {
+  }
+
+  async read() {
+    this.readState = ReadState.Reading;
+    if (!(await this._verifyFileType())) {
       this.readState = ReadState.ErrUnsupportedFileType;
       return;
     }
-    if (this.maxFileSizeByte != null && this.file.size > this.maxFileSizeByte) {
+    if (!this._verifyFileSize()) {
       this.readState = ReadState.ErrExceedMaxFileSize;
       return;
     }
-    this.readState = ReadState.Reading;
-    this.loadDimensionPromise = this._loadDimension();
+    const loadDimensionResult = await this._loadDimension();
+    if (loadDimensionResult.type === "failed") {
+      this.readState = loadDimensionResult.reason;
+      return;
+    }
+    this.readState = ReadState.ReadSuccess;
+  }
+
+  // Cache file type from header such that no need to parse again
+  __fileTypeFromHeader = null;
+  async _getFileTypeFromHeader() {
+    if (this.__fileTypeFromHeader != null) {
+      return this.__fileTypeFromHeader;
+    }
+    const arrayBuffer = await this.file.arrayBuffer();
+    const headerFileTypeBuff = new Uint8Array(arrayBuffer).subarray(0, 4);
+    let fileTypeFromHeader = "";
+    for (let i = 0; i < headerFileTypeBuff.length; i++) {
+      fileTypeFromHeader += String.fromCharCode(
+        parseInt(headerFileTypeBuff[i].toString(16), 16),
+      );
```

Comment:
sor, I am having trouble understanading this, can help explain? or maybe can help add some comment to explain 🙏